catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionContainerImpl.java

```diff
@@ -0,0 +1,396 @@
+import org.codice.ddf.catalog.subscriptionstore.internal.SubscriptionStoreException;
+import org.codice.ddf.catalog.subscriptionstore.internal.SubscriptionType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.event.Subscription;
+
+/**
+ * Container providing centralized, cached access to all {@link Subscription}s that can be registered
+ * with the system from any endpoint or subscription provider.
+ * <p>
+ * This implementation uses a {@link javax.cache.Cache} with an {@link EternalExpiryPolicy} to store
+ * instances of {@code Subscription} with the {@link org.codice.ddf.persistence.PersistentStore} as
+ * the backing store for the cache data.
+ * <p>
+ * The following "container" operations are not atomic and must be synchronized:
+ * <ul>
+ * <li>Any CRUD operation on the subscription store</li>
+ * <li>Initial loading of subscriptions from the store</li>
+ * <li>Binding and unbinding of {@link SubscriptionFactory} implementations from container clients</li>
+ * </ul>
+ */
+public class SubscriptionContainerImpl implements SubscriptionContainer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SubscriptionContainerImpl.class);
+
+    private static final String SINGLE_QUOTE = "'";
+
+    private static final String SUBSCRIPTION_CACHE_NAME = "subscription_store";
+
+    private final Cache<String, CachedSubscription> subscriptionsCache;
+
+    private final Map<String, SubscriptionFactory> factories;
+
+    public SubscriptionContainerImpl(SubscriptionCacheLoader cacheLoader,
+            SubscriptionCacheWriter cacheWriter) {
+        this.subscriptionsCache = createCache(cacheLoader, cacheWriter);
+        this.factories = createFactoryMap();
+    }
+
+    /**
+     * On startup, load existing subscriptions from the central store.
+     * <p>
+     * This is an implementation detail. We need to ensure CRUD operations, loading the cache when
+     * Karaf starts, and the binding / unbinding of factories ... are ALL mutually exclusive.
+     * <p>
+     * Event handlers to remote cache events also fit into this category. We may need to consider
+     * passing around a shared lock if this logic needs to be split out. In addition, event handlers
+     * may need to invoke this class asynchronously if they are not asynchronous themselves, otherwise
+     * we can incur deadlock by contesting the cache implementation's locking mechanism.
+     * <p>
+     * If event handlers are given this container impl, redundant writes and deletes to the backing
+     * store can <b>NOT</b> be considered errors.
+     *
+     * @see SubscriptionCacheLoader#loadAll(Iterable)
+     */
+    public synchronized void init() {
+        subscriptionsCache.loadAll(Collections.emptySet(),
+                false,
+                new CacheUpdateCompletionListener());
+
+        StreamSupport.stream(subscriptionsCache.spliterator(), false)
```

Comment:
Definitely agree here. If you had to process each entry as they are added, using the onCompletion() of the listener might be the way to go. However, doing something from that callback could also be problematic since it is called from within the cache itself.